plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphIncrementalDiagramGenerator.xtend

```diff
@@ -0,0 +1,753 @@
+            ret = toSGraph(content as KNode, context.resource.URI.toString, context.cancelIndicator)
+        }
+        return ret
+    }
+    
+    /**
+     * Translates a plain {@link KNode} or a KNode translated by {@link #translateModel} to an {@link SGraph}. 
+     * @param parentNode      the KNode that should be translated. This is the parent node containing all elements of
+     *                        the graph, that is translated
+     * @param uri             The uri of the source model the parent node was synthesized from. Used as the ID of the
+     *                        generated graph.
+     * @param cancelIndicator Indicates, if the action requesting this translation has already been canceled.
+     */
+    
+    // NOTE: This is a temporary adapter function to serve externally like the existing toSGraph function
+    // but internally enable incremental building of the SGraph
+    // this will simulate a client requesting pieces, the next steps will be incremental support in both
+    // directions (client requests and further processing such as layout not done with this sgraph though)
+    // most importantly: communication between server and client should start using incremental data
+    def SGraph toSGraph(KNode parentNode, String uri, CancelIndicator cancelIndicator) {
+        
+        kGraphToSModelElementMap = new HashMap
+        textMapping = new HashMap
+        modelLabels = new ArrayList
+        images = new HashSet
+        idGen = new KGraphElementIdGenerator
+        edgesToGenerate = new LinkedList
+        elementsToPostProcess = new LinkedList
+        childrenToProcess = new LinkedList
+        // generate an SGraph root element around the translation of the parent KNode.
+        diagramRoot = new SKGraph => [
+            type = 'graph'
+            id = uri
+            children = new ArrayList
+        ]
+
+        diagramRoot.children.addAll(incrementalCreateNodesAndPrepareEdges(parentNode, diagramRoot))
+        incrementalPostProcess()
+
+        /* ----------- Extracted to individual calls -----------------
+        // priority style queue queuing elements in viewing area first might be interesting
+        // wagon.sctx has 9 levels and 310 (node) elements
+        val maxLevel = hierarchyDepth // controls how many hierarchy levels should be generated
+        var currentLevel = 0
+        while (childrenToProcess.peek() !== null && currentLevel < maxLevel) {
+            
+            var elementsOnLevel = childrenToProcess.size()
+            while (elementsOnLevel > 0) {
+                processNextElement()
+            
+                elementsOnLevel--
+            }
+            currentLevel++
+            
+        }
+        * 
+        */
+
+        return if (cancelIndicator.canceled) 
+               null
+           else 
+               diagramRoot
+    }
+    
+    // TODO: need mechanisms to manipulate the queue in such a way to prioritize generating requested pieces
```

Comment:
Generating the SGraph here really should not take that long, if that runs parallel to the client already rendering the first pieces this should be fine already.